ompi/mca/coll/han/coll_han_scatterv.c

```diff
@@ -0,0 +1,391 @@
+        }
+
+        /* Low Scatterv */
+        low_comm->c_coll->coll_scatterv(sbuf, low_scounts, low_displs, sdtype, rbuf, rcount, rdtype,
+                                        root_low_rank, low_comm,
+                                        low_comm->c_coll->coll_scatterv_module);
+
+        char *reorder_sbuf = (char *) sbuf;
+        size_t sdsize;
+
+        ompi_datatype_type_size(sdtype, &sdsize);
+
+        up_scounts = calloc(up_size, sizeof(int));
+        up_displs = malloc(up_size * sizeof(int));
+        up_peer_ub = calloc(up_size, sizeof(int));
+        if (!up_scounts || !up_displs || !up_peer_ub) {
+            err = OMPI_ERR_OUT_OF_RESOURCE;
+            goto root_out;
+        }
+
+        for (up_peer = 0; up_peer < up_size; ++up_peer) {
+            up_displs[up_peer] = INT_MAX;
+        }
+
+        /* Calculate send counts for the inter-node scatterv */
+        for (w_peer = 0; w_peer < w_size; ++w_peer) {
+            mca_coll_han_get_ranks(vranks, w_peer, low_size, NULL, &up_peer);
+
+            if (!need_bounce_buf && root_up_rank != up_peer && 0 < scounts[w_peer] && 0 < w_peer
+                && displs[w_peer] < displs[w_peer - 1]) {
+                /* Data is not placed in the rank order so reordering is needed */
+                need_bounce_buf = 1;
+            }
+
+            if (root_up_rank == up_peer) {
+                /* No need to scatter data on the same node again */
+                continue;
+            }
+
+            up_peer_ub[up_peer] = 0 < scounts[w_peer]
+                                          && displs[w_peer] + scounts[w_peer] > up_peer_ub[up_peer]
+                                      ? displs[w_peer] + scounts[w_peer]
+                                      : up_peer_ub[up_peer];
+
+            up_scounts[up_peer] += scounts[w_peer];
+            total_up_scounts += scounts[w_peer];
+
+            /* Optimize for the happy path */
+            up_displs[up_peer] = 0 < scounts[w_peer] && displs[w_peer] < up_displs[up_peer]
+                                     ? displs[w_peer]
+                                     : up_displs[up_peer];
+        }
+
+        /* If the data is not placed contiguously on send buf, then we need a temp buf without gaps
+         */
+        for (up_peer = 0; up_peer < up_size; ++up_peer) {
+            if (root_up_rank == up_peer) {
+                continue;
+            }
+            if (!need_bounce_buf && 0 < up_scounts[up_peer]
+                && up_scounts[up_peer] != up_peer_ub[up_peer] - up_displs[up_peer]) {
+                if (up_scounts[up_peer] > up_peer_ub[up_peer] - up_displs[up_peer]) {
+                    OPAL_OUTPUT_VERBOSE(
+                        (10, mca_coll_han_component.han_output,
+                         "Detected MPI standard violation: "
+                         "The specification of counts, types, and displacements "
+                         "should not cause any location on the root to be read more than once.\n"));
```

Comment:
I opened an issue for this: https://github.com/mpi-forum/mpi-issues/issues/840

I think this warning is not necessary and I hope that the Forum removes this constraint in 4.2.

Reply:
Thank you. I removed the warning.